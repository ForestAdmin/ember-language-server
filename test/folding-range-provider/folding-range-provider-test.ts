import { createServer, getResult, ServerBucket } from './../test_helpers/public-integration-helpers';
import { MessageConnection } from 'vscode-jsonrpc/node';
import { FoldingRangeRequest } from 'vscode-languageserver-protocol/node';

describe('has folding range support', function () {
  let instance!: ServerBucket;
  let connection!: MessageConnection;
  const cursor = {
    line: 0,
    character: 0,
  };

  describe('for handlebars language', () => {
    beforeAll(async () => {
      instance = await createServer({ asyncFsEnabled: false });
      connection = instance.connection;
    });

    afterAll(async () => {
      await instance.destroy();
    });

    describe('folding range for [.hbs] files', () => {
      it('able to provide folding ranges ', async () => {
        const tpl = `
          <div>
            {{#if a}}
                hello
            {{else}}
                hello
            {{/if}}
          </div>
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('able to provide folding ranges for comments', async () => {
        const tpl = `
          
            {{!--
            {{#if a}}
                hello
            {{else}}
                hello
            {{/if}}
            --}}
    
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('able to provide folding ranges for tags', async () => {
        const tpl = `
          
            <MyComponent>

            </MyComponent>
    
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('able to provide folding ranges for blocks', async () => {
        const tpl = `
          
            {{#if}}
                a
            {{/if}}
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('able to provide folding ranges for inverse blocks', async () => {
        const tpl = `
          
            {{#if}}
                a
            {{else}}
                b
            {{/if}}
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('does not provide ranges for inline tags', async () => {
        const tpl = `
          <div></div>
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result.response.length).toBe(0);
      });
      it('does not provide ranges for inline blocks', async () => {
        const tpl = `
            {{#if a}}hello{{/if}}
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result.response.length).toBe(0);
      });
      it('does not provide ranges for inline comments', async () => {
        const tpl = `
            {{!--{{#if a}}hello{{/if}}--}}
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result.response.length).toBe(0);
      });
      it('does not fail if syntax is incorrect and return null', async () => {
        const tpl = `
          <div>
            {{#if a}}
                hello
            {{el
            {{/if}}
          </div>
        `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.hbs': tpl,
          },
          'app/components/hello/index.hbs',
          cursor
        );

        expect(result.response).toBe(null);
      });
    });

    describe('folding range for [.ts] files', () => {
      it('able to provide folding ranges ', async () => {
        const tpl = `
            import hbs from "htmlbars-inline-precompile";
            const tpl = hbs\`<div>
              {{#if a}}
                  hello
              {{else}}
                  hello
              {{/if}}
            </div>\`
          `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.ts': tpl,
          },
          'app/components/hello/index.ts',
          cursor
        );

        expect(result).toMatchSnapshot();
      });
      it('does not fail if ts syntax is incorrect and return null', async () => {
        const tpl = `
        import hbs from "htmlbars-inline-precompile";
          const tpl = hbs<div>
          {{#if a}}
              hello
          {{else}}
              hello
          {{/if}}
        </div>\`
      `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.ts': tpl,
          },
          'app/components/hello/index.ts',
          cursor
        );

        expect(result.response).toBe(null);
      });
      it('does not fail if hbs syntax is incorrect and return null', async () => {
        const tpl = `
        import hbs from "htmlbars-inline-precompile";
          const tpl = hbs\`<div>
          {{#if
              hello
          {{else}}
              hello
          {{/if}}
        </div>\`
      `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.ts': tpl,
          },
          'app/components/hello/index.ts',
          cursor
        );

        expect(result.response).toBe(null);
      });
    });

    describe('folding range for [.gts] files', () => {
      it('able to provide folding ranges ', async () => {
        const tpl = `
  
                class Component {
                    <template>
                        <div>
                            {{#if a}}
                                hello
                            {{else}}
                                hello
                            {{/if}}
                        </div>
                    </template>
                }
            `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.gts': tpl,
          },
          'app/components/hello/index.gts',
          cursor
        );

        expect(result).toMatchSnapshot();
      });

      it('does not fail if syntax is incorrect and return null', async () => {
        const tpl = `
  
        class Component {
            <template>
                <div>
                    {if a}}
                        hello
                    {{els
                        hello
                    {{/if}}
                </div>
            </template>
        }
    `.trim();
        const result = await getResult(
          FoldingRangeRequest.method,
          connection,
          {
            'app/components/hello/index.gts': tpl,
          },
          'app/components/hello/index.gts',
          cursor
        );

        expect(result.response).toBe(null);
      });
    });
  });
});
