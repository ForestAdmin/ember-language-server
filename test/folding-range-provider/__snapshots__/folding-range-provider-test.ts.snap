// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has folding range support for handlebars language folding range for [.gts] files able to provide folding ranges  1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.gts",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 31,
      "endLine": 9,
      "kind": "region",
      "startCharacter": 20,
      "startLine": 1,
    },
    Object {
      "endCharacter": 30,
      "endLine": 8,
      "kind": "region",
      "startCharacter": 24,
      "startLine": 2,
    },
    Object {
      "endCharacter": 35,
      "endLine": 7,
      "kind": "region",
      "startCharacter": 28,
      "startLine": 3,
    },
    Object {
      "endCharacter": 28,
      "endLine": 7,
      "kind": "region",
      "startCharacter": 36,
      "startLine": 5,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.hbs] files able to provide folding ranges  1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.hbs",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 16,
      "endLine": 6,
      "kind": "region",
      "startCharacter": 0,
      "startLine": 0,
    },
    Object {
      "endCharacter": 19,
      "endLine": 5,
      "kind": "region",
      "startCharacter": 12,
      "startLine": 1,
    },
    Object {
      "endCharacter": 12,
      "endLine": 5,
      "kind": "region",
      "startCharacter": 20,
      "startLine": 3,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.hbs] files able to provide folding ranges for blocks 1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.hbs",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 19,
      "endLine": 2,
      "kind": "region",
      "startCharacter": 0,
      "startLine": 0,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.hbs] files able to provide folding ranges for comments 1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.hbs",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 16,
      "endLine": 6,
      "kind": "comment",
      "startCharacter": 0,
      "startLine": 0,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.hbs] files able to provide folding ranges for inverse blocks 1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.hbs",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 19,
      "endLine": 4,
      "kind": "region",
      "startCharacter": 0,
      "startLine": 0,
    },
    Object {
      "endCharacter": 12,
      "endLine": 4,
      "kind": "region",
      "startCharacter": 20,
      "startLine": 2,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.hbs] files able to provide folding ranges for tags 1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.hbs",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 26,
      "endLine": 2,
      "kind": "region",
      "startCharacter": 0,
      "startLine": 0,
    },
  ],
}
`;

exports[`has folding range support for handlebars language folding range for [.ts] files able to provide folding ranges  1`] = `
Object {
  "addonsMeta": Array [],
  "registry": Object {
    "component": Object {
      "hello": Array [
        "app/components/hello/index.ts",
      ],
    },
  },
  "response": Array [
    Object {
      "endCharacter": 18,
      "endLine": 7,
      "kind": "region",
      "startCharacter": 28,
      "startLine": 1,
    },
    Object {
      "endCharacter": 21,
      "endLine": 6,
      "kind": "region",
      "startCharacter": 14,
      "startLine": 2,
    },
    Object {
      "endCharacter": 14,
      "endLine": 6,
      "kind": "region",
      "startCharacter": 22,
      "startLine": 4,
    },
  ],
}
`;
